File: app/src/main/java/com/itis/android/lessontwo/ui/comics/ComicsPresenter.java
package com.itis.android.lessontwo.ui.comics;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;

import com.itis.android.lessontwo.repository.RepositoryProvider;

/**
 * Created by a9 on 02.03.18.
 */
@InjectViewState
public class ComicsPresenter extends MvpPresenter<ComicsView> {

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        getViewState().getComicsId();
    }

    public void init(long id) {
        RepositoryProvider.provideComicsRepository()
                .comics(id)
                .subscribe(comics -> {
                    getViewState().setImage(comics);
                    getViewState().setDescription(comics);
                    getViewState().setPrice(comics);
                    getViewState().setPageCount(comics);
                },getViewState()::handleError);
    }
}

File: app/src/main/java/com/itis/android/lessontwo/ui/comics/ComicsPresenter.kt
package com.itis.android.lessontwo.ui.comics

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.itis.android.lessontwo.di.scope.ComicsDetailsScope
import com.itis.android.lessontwo.repository.ComicsRepository
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.addTo
import io.reactivex.rxkotlin.subscribeBy
import javax.inject.Inject

@InjectViewState
@ComicsDetailsScope
class ComicsPresenter @Inject constructor(
        private val comicsRepository: ComicsRepository
) : MvpPresenter<ComicsView>() {

    private val disposables = CompositeDisposable()

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        viewState.getComicsId()
    }

    fun init(id: Long) {
        comicsRepository.comics(id)
                .doOnSubscribe { viewState.showProgress() }
                .doAfterTerminate { viewState.hideProgress() }
                .subscribeBy(onSuccess = { comics ->
                    viewState.setImage(comics)
                    viewState.setDescription(comics)
                    viewState.setPrice(comics)
                    viewState.setPageCount(comics)
                }, onError = { viewState.handleError(it) })
                .addTo(disposables)
    }
}
